mlir/tools/mlir-tblgen/AttrOrTypeDefGen.cpp

```diff
@@ -1023,6 +1023,55 @@ bool DefGenerator::emitDefs(StringRef selectedDialect) {
   return false;
 }
 
+//===----------------------------------------------------------------------===//
+// Type Constraints
+//===----------------------------------------------------------------------===//
+
+static const char *const typeConstraintDecl = R"(
+bool {0}(::mlir::Type type);
+)";
+
+static const char *const typeConstraintDef = R"(
+bool {0}(::mlir::Type type) {
+  return ({1});
+}
+)";
```

Comment:
Small suggestion: These code blocks are so small, it's probably be more compact to just inline them into the functions that use them. Feel free to ignore though.